kernel/src/grant.rs

```diff
@@ -56,155 +27,170 @@ impl<T: ?Sized> Owned<T> {
+                    // the raw pointer.
+                    new_region.as_ptr()
+                } else if untyped_grant_ptr == (!0 as *mut u8) {
+                    return Err(Error::AlreadyInUse);
+                } else {
+                    // Grant region previously allocated, just convert the
+                    // pointer.
+                    untyped_grant_ptr as *mut T
+                };
+
+                Ok(AppliedGrant {
+                    process: process,
+                    grant_num: grant.grant_num,
+                    grant: &mut *(typed_grant_pointer as *mut T),
+                    _phantom: PhantomData,
+                })
             }
-
-            // convert `NonNull<T>` to a fat pointer `NonNull<[T]>` which includes
-            // the length information. We do this here as initialization is more
-            // convenient with the non-slice ptr.
-            let slice_ptr =
-                NonNull::new(slice_from_raw_parts_mut(ptr.as_ptr(), num_items)).unwrap();
-
-            Ok(Owned::new(slice_ptr, self.appid))
+        } else {
+            Err(Error::InactiveApp)
         }
     }
 
-    /// Like `alloc`, but the caller is responsible for free-ing the allocated
-    /// memory, as it is not wrapped in a type that implements `Drop`.
-    ///
-    /// In contrast to `alloc_raw`, this method does initialize the returned
-    /// memory.
-    unsafe fn alloc_default_unowned<T: Default>(&mut self) -> Result<NonNull<T>, Error> {
-        let ptr = self.alloc_raw()?;
-
-        // We use `ptr::write` to avoid `Drop`ping the uninitialized memory in
-        // case `T` implements the `Drop` trait.
-        write(ptr.as_ptr(), T::default());
-
-        Ok(ptr)
-    }
-
-    /// Allocates uninitialized memory appropriate to store a `T`, and returns a
-    /// pointer to said memory. The caller is responsible for both initializing the
-    /// returned memory, and dropping it properly when finished.
-    unsafe fn alloc_raw<T>(&mut self) -> Result<NonNull<T>, Error> {
-        self.alloc_n_raw::<T>(1)
-    }
-
-    /// Allocates space for a dynamic number of items. The caller is responsible
-    /// for initializing and freeing returned memory. Returns memory appropriate
-    /// for storing `num_items` contiguous instances of `T`.
-    unsafe fn alloc_n_raw<T>(&mut self, num_items: usize) -> Result<NonNull<T>, Error> {
-        let alloc_size = size_of::<T>()
-            .checked_mul(num_items)
-            .ok_or(Error::OutOfMemory)?;
-        self.appid
-            .kernel
-            .process_map_or(Err(Error::NoSuchApp), self.appid, |process| {
-                process
-                    .alloc(alloc_size, align_of::<T>())
-                    .map_or(Err(Error::OutOfMemory), |buf| {
-                        // Convert untyped `*mut u8` allocation to allocated type
-                        let ptr = NonNull::cast::<T>(buf);
-
-                        Ok(ptr)
+    fn get_if_allocated(
```

Comment:
```suggestion
    /// Return an `AppliedGrant` for a grant in a process if that
    /// grant has already been allocated.
    ///
    /// On `Err()`, returns `Err(None)` if the grant has not been allocated
    /// for this process. Returns `Err(Some(Error))` with an appropriate
    /// error otherwise.
    fn get_if_allocated(
```